### PR TITLE
SD Card driver; SD Card host drivers (SPI and SDMMC)

### PR DESCRIPTION
**Pull request by ivmarkov**
This is the `esp-idf-svc`-leg of https://github.com/esp-rs/esp-idf-hal/pull/457

Check the description of https://github.com/esp-rs/esp-idf-hal/pull/457 for more information.

### CODE REVIEW

**Comment by ivmarkov** [comment]
@AlixANNERAUD @Vollbrecht - FYI: Once the CI passes fine and I test this a bit, I'll "undraft" it - @AlixANNERAUD - you can also test this a bit if you like. You need to checkout branch "sdmmc" of `esp-idf-svc".

Comments and feedback welcome, though (and as usual) I don't promise I'll just implement all suggestions/comments. :)

**Review by Vollbrecht — COMMENTED** [review]
Overall structure looks good. 

Is the u8 drive number something we need to create a abstraction Model for? E.g making it unique to a drive and associating it to a drive "thing" ?

File: src/io.rs [hunk 1]
```diff
@@ -5,13 +5,142 @@ pub use esp_idf_hal::io::*;
...
+                sys::esp_vfs_fat_register(
+                    path.as_ptr(),
+                    drive_path.as_ptr(),
+                    max_fds as _,
+                    &mut handle,
+                )
+            })?;
+
+            unsafe {
+                sys::f_mount(handle, drive_path.as_ptr(), 0); // TODO
```
[the remark this answers is not in the document]
Comment (ivmarkov) [reply src/io.rs:120]: Thanks for noticing! Strange, these return error codes, while `ff_` don't.

Anyway, I implemented explicit checks (for now) as `f_mount` is so far the only error code we use and I `warn!` on failure and return an `ESP_FAIL` error code.

Once we expose more `f_*` APIs, we might change that.

**Comment by ivmarkov** [comment]
> Overall structure looks good.
> 
> Is the u8 drive number something we need to create a abstraction Model for? E.g making it unique to a drive and associating it to a drive "thing" ?

Yes, I thought about that but I would really like to avoid that. The cases where someone would be using more than one drive would be super-rare, so for now I think we should be fine with users having to provide the drive numbers themselves. I mean, in 99.9% of the case, there would be a single drive 0, and that's it.

**Comment by AlixANNERAUD** [comment]
Thank you for your help and contribution ^^ @ivmarkov @Vollbrecht

**Pull request merged**